lib/Dialect/FIRRTL/Transforms/ExpandWhens.cpp

```diff
@@ -103,18 +111,131 @@ class LastConnectResolver : public FIRRTLVisitor<ConcreteT> {
   using FIRRTLVisitor<ConcreteT>::visitDecl;
   using FIRRTLVisitor<ConcreteT>::visitStmt;
 
+  // This function simplifies bitindex expressions by pushing them through cat
```

Comment:
- I really hate the duplication of code here vs the canonicalizer, I hope we can find a way to 
- Same problem re: recursion on arbitrary IR as below.
- This example seems broken? (bits of bits)
```
circuit Foo:
  module Foo:
    input b: UInt<1>
    output o : UInt<2>
    wire a : UInt<2>
    a[1:0][0] <= a[1:0][1]
    a[1:0][1] <= b
    o <= a
```

Reply:
Ah yes thanks, I think we need another canonicalization for this. I'm also not a fan of the code duplication. It would be nice if we could run a canonicalization pass within expand-whens, is that possible? Another idea is that expand-whens would be three separate passes: normal expand-whens, canonicalization, initialization checking. This seems non-ideal for other reasons though.

It would really be great to run the canonicalization pass from within expand-whens, because then we fix both the code duplication and the recursion at once.

Another partial solution to the code duplication is to not add any new canonicalizations to FIRRTLFolds. It seems like these particular canonicalizations are actually applied already somewhere further down the pipeline (in HW or SV), so I don't think they actually make the output any nicer, but maybe it's still good to have them at the FIRRTL level.

Reply:
I fixed the example (now it correctly produces a combinational cycle error). It involved adding an extra bits of bits canonicalization that is also applied to the LHS, which complicates matters slightly. The LHS is guaranteed to remain a bits op after canonicalization even though the canonicalization function returns a value, so this requires a cast. Unfortunately I don't think we can enforce this guarantee through the type system unless we have two functions each for LHS and RHS canonicalization, which I think would cause some code duplication. Let me know if you think we should try to refine this more.

If we can come up with a better way to apply the canonicalizations then this whole problem may be moot.

Reply:
I investigated applying the canonicalizations directly from FIRRTLFolds, and did figure out a way to do it. Unfortunately I think the PatternRewriter API is not fine-grained enough to satisfy this use-case. Applying the BitsPrimOp canonicalizations during ExpandWhens has the side-effect of performing optimizations to uses of `bits` that were not part of subword assignments, which causes an optimization vs no optimization test to fail. Additionally, applying the canonicalization to a bits op on the LHS rewrites the connect statement but does not return the new one, so the connections stored in the driver map become invalid, leading to crashes.

After experimenting with this, I think the canonicalizations should stay in ExpandWhens and perhaps we should just not add anything to FIRRTLFolds.

Reply:
I have removed the canonicalizations from FIRRTLFolds because they are not necessary for this feature, they are duplicate code, and they cause the generated Verilog to change unnecessarily for designs that don't use bit-index.